alerting/src/main/kotlin/org/opensearch/alerting/MonitorRunner.kt

```diff
@@ -757,39 +759,119 @@ object MonitorRunner : JobRunner, CoroutineScope, AbstractLifecycleComponent() {
             return
         }
 
+        val count: Int = lastRunContext["shards_count"] as Int
+        val updatedLastRunContext = lastRunContext.toMutableMap()
+        for (i: Int in 0 until count) {
+            val shard = i.toString()
+            val maxSeqNo: Long = getMaxSeqNo(index, shard)
+            updatedLastRunContext[shard] = maxSeqNo.toString()
+        }
+
+        val queryDocIds = mutableMapOf<DocLevelQuery, Set<String>>()
+        val docsToQueries = mutableMapOf<String, MutableList<String>>()
         for (query in queries) {
-            runForEachQuery(monitor, lastRunContext, index, query)
+            val matchingDocIds = runForEachQuery(monitor, lastRunContext, updatedLastRunContext, index, query)
+            queryDocIds[query] = matchingDocIds
+            matchingDocIds.forEach {
+                if (docsToQueries.containsKey(it)) {
+                    docsToQueries[it]?.add(query.id)
+                } else {
+                    docsToQueries[it] = mutableListOf(query.id)
+                }
+            }
```

Comment:
Can we simplify this by passing a Supplier Function to ``runForEachQuery`` instead which already has the matchedDocumentIds? I am again trying to put a perspective how ``runForEachQuery`` can be executed concurrently.

Reply:
By trying to add a supplier, the code started to seem more complex and I noticed we need to change the threads the plugin runs on and additionally add additional permissions for the plugin to parallelize on multiple threads in the cluster.

We can do this as a separate PR